## 데이터셋
- 고객 성격 분석 : https://www.kaggle.com/datasets/imakash3011/customer-personality-analysis

In [1]:
import pandas as pd

In [5]:
cus = pd.read_csv('marketing_campaign.csv', sep='\t', index_col='ID')
cus

Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
ID                                                                         
5524         1957  Graduation         Single  58138.0        0         0   
2174         1954  Graduation         Single  46344.0        1         1   
4141         1965  Graduation       Together  71613.0        0         0   
6182         1984  Graduation       Together  26646.0        1         0   
5324         1981         PhD        Married  58293.0        1         0   
...           ...         ...            ...      ...      ...       ...   
10870        1967  Graduation        Married  61223.0        0         1   
4001         1946         PhD       Together  64014.0        2         1   
7270         1981  Graduation       Divorced  56981.0        0         0   
8235         1956      Master       Together  69245.0        0         1   
9405         1954         PhD        Married  52869.0        1         1   

      Dt_Customer  Recency  MntWines  MntFruits  ...  NumWebVisitsMonth  \
ID                                               ...                      
5524   04-09-2012       58       635         88  ...                  7   
2174   08-03-2014       38        11          1  ...                  5   
4141   21-08-2013       26       426         49  ...                  4   
6182   10-02-2014       26        11          4  ...                  6   
5324   19-01-2014       94       173         43  ...                  5   
...           ...      ...       ...        ...  ...                ...   
10870  13-06-2013       46       709         43  ...                  5   
4001   10-06-2014       56       406          0  ...                  7   
7270   25-01-2014       91       908         48  ...                  6   
8235   24-01-2014        8       428         30  ...                  3   
9405   15-10-2012       40        84          3  ...                  7   

       AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  \
ID                                                                            
5524              0             0             0             0             0   
2174              0             0             0             0             0   
4141              0             0             0             0             0   
6182              0             0             0             0             0   
5324              0             0             0             0             0   
...             ...           ...           ...           ...           ...   
10870             0             0             0             0             0   
4001              0             0             0             1             0   
7270              0             1             0             0             0   
8235              0             0             0             0             0   
9405              0             0             0             0             0   

       Complain  Z_CostContact  Z_Revenue  Response  
ID                                                   
5524          0              3         11         1  
2174          0              3         11         0  
4141          0              3         11         0  
6182          0              3         11         0  
5324          0              3         11         0  
...         ...            ...        ...       ...  
10870         0              3         11         0  
4001          0              3         11         0  
7270          0              3         11         0  
8235          0              3         11         0  
9405          0              3         11         1  

[2240 rows x 28 columns]

## .replace()
- https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.replace.html
- https://pandas.pydata.org/docs/reference/api/pandas.Series.replace.html
- https://pandas.pydata.org/docs/reference/api/pandas.Series.str.replace.html

In [7]:
cus['Marital_Status'].value_counts()

Marital_Status
Married     864
Together    580
Single      480
Divorced    232
Widow        77
Alone         3
Absurd        2
YOLO          2
Name: count, dtype: int64

| 결혼 상태   | 인원 수 | 뜻 |
|-----------|--------| -- |
| Married   | 864    | 기혼 |
| Together  | 580    | 동거 |
| Single    | 480    | 독신 |
| Divorced  | 232    | 이혼 |
| Widow     | 77     | 미망 |
| Alone     | 3      | 혼자, 싱글|
| Absurd    | 2      | 터무니없는 |
| YOLO      | 2      | 욜로 인생 |

이 중 Alone, Absurd, YOLO는 해석에 애매한 부분이 있다.  
해당 데이터의 토론에도 이슈가 있으며  
YOLO같은 경우는 일반적인 Single과는 다른 고객패턴이 있음을 표시하고 싶었던 것 일수도.

여기서는 이 3가지를 Single로 귀속시켜보자.

In [11]:
# 방법 1. 시리즈.replace(dict)
cus['Marital_Status'].replace({'Alone': 'Single','Absurd': 'Single','YOLO': 'Single'})
cus['Marital_Status'].replace({'Alone': 'Single','Absurd': 'Single','YOLO': 'Single'}).value_counts()

Marital_Status
Married     864
Together    580
Single      487
Divorced    232
Widow        77
Name: count, dtype: int64

In [17]:
# 방법 2. 시리즈.replace(to_replace:list, value)
cus['Marital_Status'].replace(['Alone', 'Absurd', 'YOLO'], 'Single')
cus['Marital_Status'].replace(['Alone', 'Absurd', 'YOLO'], 'Single').value_counts()

Marital_Status
Married     864
Together    580
Single      487
Divorced    232
Widow        77
Name: count, dtype: int64

다양한 replace() 예제는 아래서 살펴보자.
- https://pandas.pydata.org/docs/reference/api/pandas.Series.replace.html
- 정규표현식도 있다.

In [21]:
# 방법 3. 데이터프레임.replace() 전범위에대해 replace 를 하게됨
cus.replace({'Alone': 'Single','Absurd': 'Single','YOLO': 'Single'})

Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
ID                                                                         
5524         1957  Graduation         Single  58138.0        0         0   
2174         1954  Graduation         Single  46344.0        1         1   
4141         1965  Graduation       Together  71613.0        0         0   
6182         1984  Graduation       Together  26646.0        1         0   
5324         1981         PhD        Married  58293.0        1         0   
...           ...         ...            ...      ...      ...       ...   
10870        1967  Graduation        Married  61223.0        0         1   
4001         1946         PhD       Together  64014.0        2         1   
7270         1981  Graduation       Divorced  56981.0        0         0   
8235         1956      Master       Together  69245.0        0         1   
9405         1954         PhD        Married  52869.0        1         1   

      Dt_Customer  Recency  MntWines  MntFruits  ...  NumWebVisitsMonth  \
ID                                               ...                      
5524   04-09-2012       58       635         88  ...                  7   
2174   08-03-2014       38        11          1  ...                  5   
4141   21-08-2013       26       426         49  ...                  4   
6182   10-02-2014       26        11          4  ...                  6   
5324   19-01-2014       94       173         43  ...                  5   
...           ...      ...       ...        ...  ...                ...   
10870  13-06-2013       46       709         43  ...                  5   
4001   10-06-2014       56       406          0  ...                  7   
7270   25-01-2014       91       908         48  ...                  6   
8235   24-01-2014        8       428         30  ...                  3   
9405   15-10-2012       40        84          3  ...                  7   

       AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  \
ID                                                                            
5524              0             0             0             0             0   
2174              0             0             0             0             0   
4141              0             0             0             0             0   
6182              0             0             0             0             0   
5324              0             0             0             0             0   
...             ...           ...           ...           ...           ...   
10870             0             0             0             0             0   
4001              0             0             0             1             0   
7270              0             1             0             0             0   
8235              0             0             0             0             0   
9405              0             0             0             0             0   

       Complain  Z_CostContact  Z_Revenue  Response  
ID                                                   
5524          0              3         11         1  
2174          0              3         11         0  
4141          0              3         11         0  
6182          0              3         11         0  
5324          0              3         11         0  
...         ...            ...        ...       ...  
10870         0              3         11         0  
4001          0              3         11         0  
7270          0              3         11         0  
8235          0              3         11         0  
9405          0              3         11         1  

[2240 rows x 28 columns]

In [23]:
cus.replace({'Alone': 'Single','Absurd': 'Single','YOLO': 'Single'})['Marital_Status'].value_counts()

Marital_Status
Married     864
Together    580
Single      487
Divorced    232
Widow        77
Name: count, dtype: int64

Series.replace와 Dataframe.replace의 차이는 탐색&변경의 범위

In [25]:
# Inplcae=True로 원본에 반영하기
cus.replace({'Alone': 'Single','Absurd': 'Single','YOLO': 'Single'}, inplace=True)

## 없는 값을 바꾸려고 하면?

In [27]:
cus.replace('tmvpse', 00)

Year_Birth   Education Marital_Status   Income  Kidhome  Teenhome  \
ID                                                                         
5524         1957  Graduation         Single  58138.0        0         0   
2174         1954  Graduation         Single  46344.0        1         1   
4141         1965  Graduation       Together  71613.0        0         0   
6182         1984  Graduation       Together  26646.0        1         0   
5324         1981         PhD        Married  58293.0        1         0   
...           ...         ...            ...      ...      ...       ...   
10870        1967  Graduation        Married  61223.0        0         1   
4001         1946         PhD       Together  64014.0        2         1   
7270         1981  Graduation       Divorced  56981.0        0         0   
8235         1956      Master       Together  69245.0        0         1   
9405         1954         PhD        Married  52869.0        1         1   

      Dt_Customer  Recency  MntWines  MntFruits  ...  NumWebVisitsMonth  \
ID                                               ...                      
5524   04-09-2012       58       635         88  ...                  7   
2174   08-03-2014       38        11          1  ...                  5   
4141   21-08-2013       26       426         49  ...                  4   
6182   10-02-2014       26        11          4  ...                  6   
5324   19-01-2014       94       173         43  ...                  5   
...           ...      ...       ...        ...  ...                ...   
10870  13-06-2013       46       709         43  ...                  5   
4001   10-06-2014       56       406          0  ...                  7   
7270   25-01-2014       91       908         48  ...                  6   
8235   24-01-2014        8       428         30  ...                  3   
9405   15-10-2012       40        84          3  ...                  7   

       AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  AcceptedCmp2  \
ID                                                                            
5524              0             0             0             0             0   
2174              0             0             0             0             0   
4141              0             0             0             0             0   
6182              0             0             0             0             0   
5324              0             0             0             0             0   
...             ...           ...           ...           ...           ...   
10870             0             0             0             0             0   
4001              0             0             0             1             0   
7270              0             1             0             0             0   
8235              0             0             0             0             0   
9405              0             0             0             0             0   

       Complain  Z_CostContact  Z_Revenue  Response  
ID                                                   
5524          0              3         11         1  
2174          0              3         11         0  
4141          0              3         11         0  
6182          0              3         11         0  
5324          0              3         11         0  
...         ...            ...        ...       ...  
10870         0              3         11         0  
4001          0              3         11         0  
7270          0              3         11         0  
8235          0              3         11         0  
9405          0              3         11         1  

[2240 rows x 28 columns]

아무 에러도 발생하지 않는다.